core/trino-main/src/main/java/io/trino/operator/output/SkewedPartitionRebalancer.java

```diff
@@ -0,0 +1,412 @@
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.ceil;
+import static java.lang.Math.floorMod;
+import static java.lang.Math.max;
+
+/**
+ * Helps in distributing big or skewed partitions across available tasks to improve the performance of
+ * partitioned writes.
+ * <p>
+ * This rebalancer initialize a bunch of buckets for each task based on a given taskBucketCount and then tries to
+ * uniformly distribute partitions across those buckets. This helps to mitigate two problems:
+ * 1. Mitigate skewness across tasks.
+ * 2. Scale few big partitions across tasks even if there's no skewness among them. This will essentially speed the
+ *    local scaling without impacting much overall resource utilization.
+ * <p>
+ * Example:
+ * <p>
+ * Before: 3 tasks, 3 buckets per task, and 2 skewed partitions
+ * Task1                Task2               Task3
+ * Bucket1 (Part 1)     Bucket1 (Part 2)    Bucket1
+ * Bucket2              Bucket2             Bucket2
+ * Bucket3              Bucket3             Bucket3
+ * <p>
+ * After rebalancing:
+ * Task1                Task2               Task3
+ * Bucket1 (Part 1)     Bucket1 (Part 2)    Bucket1 (Part 1)
+ * Bucket2 (Part 2)     Bucket2 (Part 1)    Bucket2 (Part 2)
+ * Bucket3              Bucket3             Bucket3
+ */
+@ThreadSafe
+public class SkewedPartitionRebalancer
+{
+    private static final Logger log = Logger.get(SkewedPartitionRebalancer.class);
+    // Keep the scale writers partition count big enough such that we could rebalance skewed partitions
+    // at more granularity, thus leading to less resource utilization at writer stage.
+    private static final int SCALE_WRITERS_PARTITION_COUNT = 4096;
+    // If the percentage difference between the two different task buckets with maximum and minimum processed bytes
+    // since last rebalance is above 0.7 (or 70%), then we consider them skewed.
+    private static final double TASK_BUCKET_SKEWNESS_THRESHOLD = 0.7;
+    private static final long MIN_DATA_PROCESSED_REBALANCE_THRESHOLD = DataSize.of(50, MEGABYTE).toBytes();
+
+    private final int partitionCount;
+    private final int taskCount;
+    private final int taskBucketCount;
+    private final long minPartitionDataProcessedRebalanceThreshold;
+    private final long minDataProcessedRebalanceThreshold;
+
+    private final AtomicLongArray partitionRowCount;
+    private final AtomicLong dataProcessed;
+    private final AtomicLong dataProcessedAtLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionDataSizeAtLastRebalance;
+
+    @GuardedBy("this")
+    private final long[] partitionDataSizeSinceLastRebalancePerTask;
+
+    @GuardedBy("this")
+    private final long[] estimatedTaskBucketDataSizeSinceLastRebalance;
+
+    private final List<List<TaskBucket>> partitionAssignments;
+
+    public static PartitionFunction createPartitionFunction(
```

Comment:
I think we need to avoid global scaling when `ConnectorBucketNodeMap#hasFixedMapping` is true.

I would say that `ConnectorTableLayout#multipleWritersPerPartitionSupported=true` and `ConnectorBucketNodeMap#hasFixedMapping=true` is legal. E.g. partition needs to be written by a given node, but locally on a node it might be scaled